File: Zamba/zambawindows2013/Zamba2010/Zamba.Bussines/Security/UserPreferences.vb
Imports Zamba.Tools
Imports Zamba.Data

Public Class UserPreferences
   
    '''Devuelve el valor de un item por el nombre
    Public Shared Function getValue(ByVal name As String, ByVal Section As Sections, ByVal DefaultValue As Object) As String
        Try
            If Cache.UsersAndGroups.hsUserPreferences.Contains(name) Then
                Return Cache.UsersAndGroups.hsUserPreferences(name)
            Else
                Dim valor As String
                Dim userId As Int64

                'Si no hay un usuario logueado, busco en la tabla de PC's zmachineconfig
                If Membership.MembershipHelper.CurrentUser Is Nothing Or name.ToLower() = "userid" Then
                    'busco el valor especifico de la PC
                    valor = UserPreferencesFactory.getValueDBByMachine(name, Section, Environment.MachineName)

                    If valor Is Nothing Then
                        'si no hay valor especifico de la PC busco el valor por default para las PC
                        valor = UserPreferencesFactory.getValueDBByMachine(name, Section, "default")

                        'Si no esta en la zmachineconfig el valor por defecto de las PC
                        If valor Is Nothing Then
                            'Guardo el valor por defecto del codigo en el valor por defecto de las PC
                            UserPreferencesFactory.setValueDBByMachine(name, DefaultValue, Section, "default")
                            valor = DefaultValue
                        End If
                    End If
                Else
                    'Hay usuario logueado, entonces vamos a usar la zuserconfig
                    userId = Membership.MembershipHelper.CurrentUser.ID
                    'busco el valor especifico del usuario
                    valor = UserPreferencesFactory.getValueDB(name, Section, userId)

                    If valor Is Nothing Then
                        'Si no hay valor especifico del usuario  busco el valor por defecto Usuario 0
                        valor = UserPreferencesFactory.GetDefaultValueDB(name, Section)

                        'Si no tiene un valor por defecto de usuario 0 en la base
                        If valor Is Nothing Then
                            'guardo el valor por defecto del codigo en el usuario por defecto 0
                            UserPreferencesFactory.setValueDB(name, DefaultValue, Section, 0)
                            valor = DefaultValue
                        End If
                    End If
                End If

                If Not Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name) Then
                    Cache.UsersAndGroups.hsUserPreferences.Add(name, valor)
                End If

                Return valor
            End If

        Catch ex As System.Threading.ThreadAbortException
        Catch ex As Exception
            Zamba.Core.ZClass.raiseerror(New Exception("Error al buscar el valor de " & name & " " & ex.ToString))
            Return DefaultValue
        End Try
    End Function

    '''Devuelve el valor de un item por el nombre
    Public Shared Function getValueForMachine(ByVal name As String, ByVal Section As Sections, ByVal DefaultValue As Object) As String
        Try
            If Cache.UsersAndGroups.hsUserPreferences.Contains(name) Then
                Return Cache.UsersAndGroups.hsUserPreferences(name)
            Else
                Dim valor As String = UserPreferencesFactory.getValueDBByMachine(name, Section, Environment.MachineName)

                If valor Is Nothing Then
                    'si no hay valor especifico de la PC busco el valor por default para las PC
                    valor = UserPreferencesFactory.getValueDBByMachine(name, Section, "default")

                    'Si no esta en la zmachineconfig el valor por defecto de las PC
                    If valor Is Nothing Then
                        'Guardo el valor por defecto del codigo en el valor por defecto de las PC
                        UserPreferencesFactory.setValueDBByMachine(name, DefaultValue, Section, "default")
                        valor = DefaultValue
                    End If
                End If

                If Not Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name) Then
                    Cache.UsersAndGroups.hsUserPreferences.Add(name, valor)
                End If

                Return valor
            End If
        Catch ex As System.Threading.ThreadAbortException
        Catch ex As Exception
            Zamba.Core.ZClass.raiseerror(New Exception("Error al buscar el valor de " & name & " " & ex.ToString))
            Return DefaultValue
        End Try
    End Function

    '''agrega un item por el nombre y le asigna el valor
    Public Shared Sub setValue(ByVal name As String, ByVal valor As String, ByVal Section As Sections)
        Try
            'Si no hay un usuario logueado, busco en la tabla de PC's
            If Membership.MembershipHelper.CurrentUser Is Nothing Then
                UserPreferencesFactory.setValueDBByMachine(name, valor, Section, Environment.MachineName)
            Else
                UserPreferencesFactory.setValueDB(name, valor, Section, Membership.MembershipHelper.CurrentUser.ID)
            End If
            If (Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name)) Then
                Cache.UsersAndGroups.hsUserPreferences(name) = valor
            Else
                Cache.UsersAndGroups.hsUserPreferences.Add(name, valor)
            End If
        Catch ex As Exception
            Zamba.Core.ZClass.raiseerror(ex)
        End Try
    End Sub

    '''agrega un item por el nombre y le asigna el valor
    Public Shared Sub setValueForMachine(ByVal name As String, ByVal valor As String, ByVal Section As Sections)
        Try
            UserPreferencesFactory.setValueDBByMachine(name, valor, Section, Environment.MachineName)
            If (Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name)) Then
                Cache.UsersAndGroups.hsUserPreferences(name) = valor
            Else
                Cache.UsersAndGroups.hsUserPreferences.Add(name, valor)
            End If
        Catch ex As Exception
            Zamba.Core.ZClass.raiseerror(ex)
        End Try
    End Sub

    Enum InitialSizes
        Normal
        Height
        Width
    End Enum

    Public Shared Sub LoadAllMachineConfigValues()
        Dim values As DataTable = Nothing
        Try
            values = UserPreferencesFactory.getAllValuesDBByMachine(Environment.MachineName)
            SyncLock Cache.UsersAndGroups.hsUserPreferences
            For Each Value As DataRow In values.Rows
                If Cache.UsersAndGroups.hsUserPreferences.ContainsKey(Value(1)) = False Then Cache.UsersAndGroups.hsUserPreferences.Add(Value(1), Value(0))
            Next
            End SyncLock
        Catch ex As Exception
            Zamba.Core.ZClass.raiseerror(New Exception("Error al cargar las preferencias de maquina" & ex.ToString))
        Finally
            If values IsNot Nothing Then
                values.Dispose()
                values = Nothing
            End If
        End Try
    End Sub

    Public Shared Sub LoadAllUserConfigValues()
        If Membership.MembershipHelper.CurrentUser IsNot Nothing Then
            Dim values As DataTable = Nothing
            Try
                values = UserPreferencesFactory.getAllValuesDB(Membership.MembershipHelper.CurrentUser.ID)
                SyncLock Cache.UsersAndGroups.hsUserPreferences
                For Each Value As DataRow In values.Rows
                    If Cache.UsersAndGroups.hsUserPreferences.ContainsKey(Value("name")) = False Then Cache.UsersAndGroups.hsUserPreferences.Add(Value("name"), Value("value"))
                Next
                    End SyncLock
            Catch ex As Exception
                Zamba.Core.ZClass.raiseerror(New Exception("Error al cargar las preferencias del usuario" & ex.ToString))
            Finally
                If values IsNot Nothing Then
                    values.Dispose()
                    values = Nothing
                End If
            End Try
        End If
    End Sub

End Class

File: Zamba/Zamba.Web/Zamba.FUS/UserPreferences.cs
using System;
using System.Data;
using Zamba;
using Zamba.Servers;

public partial class UserPreferences
{
    public enum InitialSizes
    {
        Normal,
        Height,
        Width
    }

    ///Devuelve el valor de un item por el nombre
    public  string getValue(string name, UPSections Section, object DefaultValue)
    {
        try
        {
            if (Server.ConInitialized || Server.ConInitializing)
            {
                if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Contains(name))
                {
                    if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name] != null)
                        return Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name].ToString();
                    else
                        return string.Empty;
                }
                else
                {
                    string valor = null;

                    //Si no hay un usuario logueado, busco en la tabla de PC's zmachineconfig
                    if (Zamba.Membership.MembershipHelper.CurrentUser == null | name.ToLower() == "userid")
                    {
                        //busco el valor especifico de la PC
                        valor = UserPreferencesFactory.getValueDBByMachine(name, Section, Environment.MachineName);

                        if (valor == null)
                        {
                            //si no hay valor especifico de la PC busco el valor por default para las PC
                            valor = UserPreferencesFactory.getValueDBByMachine(name, Section, "default");

                            //Si no esta en la zmachineconfig el valor por defecto de las PC
                            if (valor == null)
                            {
                                //Guardo el valor por defecto del codigo en el valor por defecto de las PC
                                UserPreferencesFactory.setValueDBByMachine(name, DefaultValue.ToString(), Section, "default");
                                valor = DefaultValue.ToString();
                            }
                        }
                    }
                    else
                    {
                        //Hay usuario logueado, entonces vamos a usar la zuserconfig

                        //busco el valor especifico del usuario
                        valor = UserPreferencesFactory.getValueDB(name, Section, Zamba.Membership.MembershipHelper.CurrentUser.ID);

                        if (valor == null)
                        {
                            //Si no hay valor especifico del usuario  busco el valor por defecto Usuario 0
                            valor = UserPreferencesFactory.GetDefaultValueDB(name, Section);

                            //Si no tiene un valor por defecto de usuario 0 en la base
                            if (valor == null)
                            {
                                //guardo el valor por defecto del codigo en el usuario por defecto 0
                                UserPreferencesFactory.setValueDB(name, DefaultValue.ToString(), Section, 0);
                                valor = DefaultValue.ToString();
                            }
                        }
                    }

                    if (!Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name))
                    {
                        Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(name, valor);
                    }

                    return valor;
                }
            }
            else
            {
                return DefaultValue.ToString();
            }
        }
        catch (System.Threading.ThreadAbortException)
        {
            return DefaultValue.ToString();
        }
        catch (ZConnectionException ex)
        {
            throw ex;
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(new Exception("Error al buscar el valor de " + name + " " + ex.ToString()));
            return DefaultValue.ToString();
        }
    }


    ///Devuelve el valor de un item por el nombre
    public static string getValueForMachine(string name, UPSections Section, object DefaultValue, bool UseCache = true)
    {
        return getValueForMachine(name, Section, DefaultValue, Environment.MachineName, UseCache);
    }

    public static string getValueForMachine(string name, UPSections Section, object DefaultValue, string machineName, bool UseCache = true)
    {
        try
        {
            if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Contains(name) && UseCache)
            {
                if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name] != null)
                    return Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name].ToString();
                else
                    return string.Empty;
            }
            else
            {
                string valor = UserPreferencesFactory.getValueDBByMachine(name, Section, machineName);

                if (valor == null)
                {
                    //si no hay valor especifico de la PC busco el valor por default para las PC
                    valor = UserPreferencesFactory.getValueDBByMachine(name, Section, "default");

                    //Si no esta en la zmachineconfig el valor por defecto de las PC
                    if (valor == null)
                    {
                        //Guardo el valor por defecto del codigo en el valor por defecto de las PC
                        UserPreferencesFactory.setValueDBByMachine(name, DefaultValue.ToString(), Section, "default");
                        valor = DefaultValue.ToString();
                    }
                }

                if (!Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name))
                {

                    Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(name, valor);
                }
                else if (UseCache == false)
                    Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name] = valor;


                return valor;
            }
        }
        catch (System.Threading.ThreadAbortException)
        {
            return DefaultValue.ToString();
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(new Exception("Error al buscar el valor de " + name + " " + ex.ToString()));
            return DefaultValue.ToString();
        }
    }

    ///agrega un item por el nombre y le asigna el valor
    public static void setValue(string name, string valor, UPSections Section)
    {
        try
        {
            //Si no hay un usuario logueado, busco en la tabla de PC's
            if (Zamba.Membership.MembershipHelper.CurrentUser == null)
            {
                UserPreferencesFactory.setValueDBByMachine(name, valor, Section, Environment.MachineName);
            }
            else
            {
                UserPreferencesFactory.setValueDB(name, valor, Section, Zamba.Membership.MembershipHelper.CurrentUser.ID);
            }
            if ((Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name)))
            {
                Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name] = valor;
            }
            else
            {
                Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(name, valor);
            }
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(ex);
        }
    }

    ///agrega un item por el nombre y le asigna el valor
    public static void setValueForMachine(string name, string valor, UPSections Section)
    {
        setValueForMachine(name, valor, Section, Environment.MachineName);
    }

    public static void setValueForMachine(string name, string valor, UPSections Section, string MachineName)
    {
        try
        {
            UserPreferencesFactory.setValueDBByMachine(name, valor, Section, MachineName);
            if ((Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(name)))
            {
                Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[name] = valor;
            }
            else
            {
                Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(name, valor);
            }
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(ex);
        }
    }

    /// agrega un item por el nombre y le asigna el valor para un userId especificado.
    public static void setValueForUsersId(string name, string value, UPSections section, long userId)
    {
        UserPreferencesFactory.setValueDB(name, value, section, userId);
    }


    public static void LoadAllMachineConfigValues()
    {
        DataTable values = null;
        try
        {
            values = UserPreferencesFactory.getAllValuesDBByMachine(Environment.MachineName);
            lock (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences)
            {
                foreach (DataRow Value in values.Rows)
                {
                    if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(Value["name"]) == false)
                        Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(Value["name"], Value["value"]);
                }
            }
        }
        catch (Exception ex)
        {
            Zamba.AppBlock.ZException.Log(new Exception("Error al cargar las preferencias de maquina" + ex.ToString()));
        }
        finally
        {
            if (values != null)
            {
                values.Dispose();
                values = null;
            }
        }
    }
    public static void LoadAllUserConfigValues()
    {
        if (Zamba.Membership.MembershipHelper.CurrentUser != null)
        {
            DataTable values = null;
            try
            {
                values = UserPreferencesFactory.getAllValuesDB(Zamba.Membership.MembershipHelper.CurrentUser.ID);
                lock (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences)
                {
                    foreach (DataRow Value in values.Rows)
                    {
                        if (Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.ContainsKey(Value["name"]) == false)
                            Zamba.Core.Cache.UsersAndGroups.hsUserPreferences.Add(Value["name"], Value["value"]);
                        else
                            Zamba.Core.Cache.UsersAndGroups.hsUserPreferences[Value["name"]] = Value["value"];
                    }
                }
            }
            catch (Exception ex)
            {
                Zamba.AppBlock.ZException.Log(new Exception("Error al cargar las preferencias del usuario" + ex.ToString()));
            }
            finally
            {
                if (values != null)
                {
                    values.Dispose();
                    values = null;
                }
            }
        }
    }

    public static void RemoveValueByUsersID(string userID, string settingName)
    {
        UserPreferencesFactory.DeleteUserSetting(userID, settingName);
    }
}